### PR TITLE
PP-719 Resource to create or update notification credentials

### PR DESCRIPTION
**Pull request by imranbohoran**
## WHAT

_A brief description of the pull request:_
- Introduced a HashUtil to performa a one-way hash on a string
- Introduced an EntityBuilder that creates a new instance of NotificationCredentials.
  This makes it easy to unit test the resource as it's doing quite a few things.
- The resource expects to receive a json payload which has username and password fields.
- The password field size has been changed as Bcrypt hashed string is more than 25 characters.

with @mrlumbu

### CODE REVIEW

**Comment by imranbohoran** [comment]
rebuild

**Comment by imranbohoran** [comment]
rebuild

**Comment by imranbohoran** [comment]
rebuild

**Comment by kakumara** [comment]
just observation: favour few commits with each doing a unit of work over one chunky commit does many things.

**Comment by imranbohoran** [comment]
👍  We could have broken it into create and update separately. Most of the changes are actually the unit and integration tests. We had to do a bit more to get some proper unit tests written as we don't seem to have enough of those

File: src/main/java/uk/gov/pay/connector/model/builder/EntityBuilder.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.pay.connector.model.builder;
+
+import uk.gov.pay.connector.model.domain.GatewayAccountEntity;
+import uk.gov.pay.connector.model.domain.NotificationCredentials;
+
+public class EntityBuilder {
+
+    public NotificationCredentials newNotificationCredentials(GatewayAccountEntity gatewayAccountEntity) {
+        return new NotificationCredentials(gatewayAccountEntity);
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/connector/model/builder/EntityBuilder.java]: Its a one liner thats been wrapped by this EntityBuilder. Cannot see the benefit of using this as you have a constructor doing the job

Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/model/builder/EntityBuilder.java]: We extracted this out so we can have some proper unit tests and run assertions on the object being created.

File: src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java [hunk 2]
```diff
@@ -227,6 +236,38 @@ public Response updateGatewayAccountAcceptedCardTypes(@PathParam("accountId") Lo
                         notFoundResponse(format("The gateway account id '%s' does not exist", gatewayAccountId)));
     }
 
+    @POST
+    @Path(GATEWAY_ACCOUNTS_NOTIFICATION_CREDENTIALS)
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON)
+    @Transactional
+    public Response createOrUpdateGatewayAccountNotificationCredentials(@PathParam("accountId") Long gatewayAccountId, Map<String,String> notificationCredentials) {
+        if (!notificationCredentials.containsKey(USERNAME_KEY)) {
```
Comment (kakumara) [inline src/main/java/uk/gov/pay/connector/resources/GatewayAccountResource.java]: resource method doing bit too much perhaps? can see doing 3 things validation, doing the job, handling response. possible refactoring?

**Comment by kakumara** [comment]
other than that 👍

**Comment by rahul-vinayak** [comment]
rebuild

**Comment by mrlumbu** [comment]
rebuild

**Comment by imranbohoran** [comment]
rebuild

**Comment by ijmad** [comment]
rebuild

**Comment by ijmad** [comment]
rebuild

**Comment by ijmad** [comment]
rebuild

**Pull request merged**